### ISSUE TITLE
compiler: fp8-blockwise GEMV linear (`gemv_fp8`) — dequant-fused decode projection

### ISSUE THREAD

**Issue by xzyaoi** [opentela-ai/vkernels#91]
## Context
Decode-step projections with DeepSeek-style 128×128 block-FP8 weights (qwen35 27B checkpoints, deepseek_v41 `Fp8Linear`, GLM expert weights MXFP4 — see follow-up note).

## Semantics
`y = x @ dequant(w_fp8, scales[K/128, N/128])` — numerics: fp32 accumulation over K, dequant in-register per block; scales fp32.

## Work
- `linear` op variant / attributes: `weight_layout: "fp8_block"`, scale tensor as second weight external
- lowering: same GEMV tile domain as `linear`; scales region reads
- device template **already implemented** (`_h_gemv_fp8` in `device_triton_hybrid.py`, validated at 29.7 GB resident fp8 weights on GB10)
- reference body: numpy dequant-then-matmul in fp64
- MXFP4 expert weights (glm/deepseek checkpoints) are a dtype follow-up on the same task shape

## Validation
bf16-dequantized-reference vs fp8 GEMV rel-err bound, mirroring the 27B gate (logits 1.4–3.4e-3).

Labels: arch:ops, model:qwen3.5, model:deepseek-v4, model:glm-5.3

## Implementation sketch

**Backend**: Triton. Template **already validated** as `_h_gemv_fp8` (`device_triton_hybrid.py`, 27B, 29.7 GB fp8-resident): one task per 16-column tile of N, full-K sweep with 128-deep K blocks.

**Arithmetic** (dequant-in-register, fp32 accumulate):
```
for kb in K/128:  s = scales[sb_row, kb]          # one fp32 per 128×128 block
    acc += Σ_k w_fp8[n, k] · (s · x[k])           # e4m3 → fp32 in-register
y[n] = acc                                          # k ascending (contract)
```
**Dtypes**: weights **e4m3** `[N, K]` row-major; block scales **fp32** `[N/128 · K/128]` block-major (per-layer stacked, layer offset at runtime); x bf16; y bf16 workspace. Memory-bound GEMV at B≤8 — no `tl.dot` needed; revisit tiling if B grows.
**MXFP4 follow-up** (glm/deepseek checkpoints): same task shape, 4-bit weights + block-32 e8m0 group scales — separate dequant body only.

**Comment by xzyaoi** [comment]
## Status: implemented — compiler vertical slice landed on `issue-91-gemv-fp8` (3 commits, PR to follow)

CPU compiler lane only (no GPU/Slurm jobs used). Device template untouched, per the issue: `_t_gemv_fp8` (`device_triton.py`) / `_h_gemv_fp8` (`device_triton_hybrid.py`) remain the validated device bodies; the compiler now lowers to them.

### What landed
- **IR** (`operator_ir.py`): `OP_LINEAR_FP8` op kind + `F8_E4M3` element dtype (1 byte).
- **Capture** (`capture.py`): `ops.linear_fp8(x, w_fp8, scale)` — the `linear` variant recording `weight_layout="fp8_block"`, `quant_block=128`, scale tensor as the **second weight external**; validates `K % 128`, `N % 16` (ragged trailing scale block allowed, `ceil(N/128)` scale rows), contraction match, and the block-major scale shape; numerical contract records per-block dequant + f32 k-ascending accumulation.
- **Lowering** (`lowerings/`): `lower_linear_fp8` → family kind `gemv_fp8`, **same GEMV tile domain as `linear`** (`(m,16) × (n,16)` tiles, full-K sweep). Per-task region reads: the 16×K weight tile, the m×K activation slice, and **exactly one scale row** (a 16-column tile always lies inside one 128-wide N block) + one fp32 scale per 128-deep k-block. Registered in `LOWERINGS`; `TEMPLATE_NAMES["gemv_fp8"] = "linear_fp8_task"` for the codegen phase table.
- **Reference body** (`reference_exec.py`): `_body_gemv_fp8` — numpy dequant-then-matmul in fp64, tile-exact per-block scale broadcast; plus a vectorized `decode_e4m3` (e4m3fn → fp64, subnormals + NaN-0x7F handled) used by the oracle suite.
- **Docs**: compiler README section for the fp8-blockwise GEMV lane.

### Test evidence (`tests/python/test_linear_fp8_compiler.py`, 17 tests, all CPU, <1s)
- reference-executor output vs independent fp64 dequant-matmul oracle: **max rel err < 1e-12**, NaN-canary clean (every tile written exactly once).
- device-template mirror (numpy simulation of `_t_gemv_fp8`'s exact decomposition: 16-col tiles, 128-deep k blocks, one fp32 scale load per block, fp32 accumulate, worker-stride assignment) vs reference executor: **< 1e-5** across `workers ∈ {1,3,7,13}`, odd m, ragged N=144, single- and multi-block K, block-boundary tiles (n0=112 vs 128 → scale rows 0 vs 1).
- capture/lowering unit cases: layout attribute, input order `(x, w, scale)`, scale-region shape per task, rejection of bad scale shape / contraction mismatch / non-128 K / wrong dtypes / wrong `weight_layout`.
- fp8-vs-bf16-reference bound (issue's validation item): measured **2.6e-2 / 3.0e-2 / 3.1e-2** max rel err over 3 seeds for a single K=512 GEMV — bound pinned at 5e-2 with the rationale documented (the 27B gate's 1.4–3.4e-3 is the end-to-end 64-layer logits regime; per-GEMV e4m3 error is the term measured here).
- Regression: full `tests/python` run — 309 passed, 172 skipped (GPU-gated), 1 pre-existing failure (`test_discovery.py::test_kernel_names_and_order`, fails identically with the change stashed at base 4c58c58).

### Remaining risk
- **Device-side end-to-end**: the compiled `gemv_fp8` family is validated against the template *algorithm* on CPU (§15.1); wiring a compiled fp8 schedule into a live Triton launch (e.g. a deepseek_v41 frontend feeding `TritonMegakernel`) is not exercised here — no GPU in this lane.
- **Ragged N** (N%128≠0) is accepted per the template's constraint (N%16), but real checkpoints are 128-aligned; the ragged path is CPU-oracle-tested only.
- **MXFP4 expert weights** (glm/deepseek) remain the dtype follow-up noted in the issue — same task shape, separate dequant body.
- `y` is fp32 (workspace default); bf16 workspace output would match the hybrid kernel's store dtype — trivial follow-up if the frontend demands it.
